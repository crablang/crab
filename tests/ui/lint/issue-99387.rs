//! Test that we don't follow through projections to find
//! opaque types.

#![feature(type_alias_impl_trait)]
#![allow(private_in_public)]

pub type Successors<'a> = impl Iterator<Item = &'a ()>;

pub fn f<'a>() -> Successors<'a> {
    None.into_iter()
}

trait Tr {
    type Item;
}

impl<'a> Tr for &'a () {
    type Item = Successors<'a>;
}

pub fn ohno<'a>() -> <&'a () as Tr>::Item {
    //~^ ERROR item constrains opaque type that is not in its signature
    None.into_iter()
}

fn main() {}
