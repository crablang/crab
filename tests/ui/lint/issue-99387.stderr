error: item constrains opaque type that is not in its signature
  --> $DIR/issue-99387.rs:21:22
   |
LL | pub fn ohno<'a>() -> <&'a () as Tr>::Item {
   |                      ^^^^^^^^^^^^^^^^^^^^
   |
   = note: this item must mention the opaque type in its signature in order to be able to register hidden types
note: this item must mention the opaque type in its signature in order to be able to register hidden types
  --> $DIR/issue-99387.rs:21:8
   |
LL | pub fn ohno<'a>() -> <&'a () as Tr>::Item {
   |        ^^^^

error: aborting due to previous error

